slides/aml-10-021918-trees-forests/index.html

```diff
@@ -0,0 +1,724 @@
+tree_dot = export_graphviz(tree,out_file=
+None,feature_names=iris.feature_names)
+graphviz.Source(tree_dot)
+```
+]
+![:scale 28%](images/img_22.png)
+]
+???
+---
+
+class: split-40
+# Feature importance
+
+.column[
+.smaller[```python
+X_train, X_test, y_train, y_test = 
+train_test_split(iris.data, iris.target, stratify=iris.target, random_state=1)
+tree = DecisionTreeClassifier(max_leaf_no
+des=6).fit(X_train,y_train)
+tree_dot = export_graphviz(tree,out_file=
+None,feature_names=iris.feature_names)
+graphviz.Source(tree_dot)
+```]
+![:scale 50%](images/img_22.png)
+]
+
+.column[
+
+.smaller[
+```python
+tree.feature_importances_
+array([0.0, 0.0, 0.414, 0.586])
+```
+]
+
+![:scale 50%](images/img_23.png)
+
+.smaller[
+- Unstable Tree -> Unstable feature importances.
+- Might take one or multiple from a group of correlated features.
+]
+]
+
+???
+---
+
+# Categorical Data
+
+- Can split on categorical data directly
+- Intuitive way to split: split in two subsets
+- 2 ^ n_values many possibilities
+- Possible to do in linear time exactly for gini index and binary classification.
+- Heuristics done in practice for multi-class.
+- Not in sklearn release version :(
+
+???
+
+---
+class:spacious
+# Predicting probabilities
+- Fraction of class in leaf.
+- Without pruning: Always 100% certain!
+- Even with pruning might be too certain.
+???
+---
+
+class:spacious
+# Conditional Inference Trees
+
+- Select “best” split with correcting for multiple-hypothesis testing.
+- More “fair” to categorical variables.
+- Only in R so far (party)
+???
+---
+
+class:spacious
+# Relation to Nearest Neighbors
+- Predict average of neighbors – either by k, by epsilon ball or by leaf.
+- Trees are much faster to predict.
+- Both can’t extrapolate
+???
+---
+
+# Different splitting methods
+- Could use anything as split candidate!
+- Linear models used if extrapolation is needed.
+- Computer vision: pixel comparisons
+- Kinect (first generation): depth comparison
+
+.center[![:scale 50%](images/img_24.png)]
+
+.smaller[(taken from Shotton et. al. Real-Time Human Pose Recognition ..)]
+
+???
+---
+class: centre,middle
+# Ensemble Models
+???
+---
+
+
+# Poor man’s ensembles
+
+- Build different models
+- Average the result
+- Owen Zhang (long time kaggle 1st): build XGBoosting models with different random seeds.
+- More models are better – if they are not correlated.
+- Also works with neural networks
+- You can average any models as long as they provide calibrated (“good”) probabilities.
+- Scikit-learn: VotingClassifier hard and soft voting
+
+???
+---
+
+# VotingClassifier
+
+.smaller[
+```python
+voting = VotingClassifier([('logreg', LogisticRegression(C=100)),
+('tree', DecisionTreeClassifier(max_depth=3, random_state=0))],
+voting='soft')
+voting.fit(X_train, y_train)
+lr, tree = voting.estimators_
+print(("{:.2f} " * 3).format(voting.score(X_test, y_test),
+lr.score(X_test, y_test), tree.score(X_test, y_test)))
+```
+0.88 0.84 0.80 
```

Comment:
output should have triple backticks around it.

Reply:
Resolved